examples/bond/assets/bond-dapp-stateful.py

```diff
@@ -0,0 +1,209 @@
+        Return(Int(1))
+    ])
+
+    # store buyback address
+    create_buyback = Seq([
+        Assert(
+            And(
+                Txn.sender() == App.globalGet(store_manager),
+                basic_checks
+            )
+        ),
+        App.globalPut(Bytes("buyback"), Txn.application_args[1]),
+        Return(Int(1))
+    ])
+
+    on_exit = Seq([
+        Assert(
+            And(
+                basic_checks,
+                # verify first transaction is asset transfer
+                Gtxn[0].type_enum() == TxnType.AssetTransfer,
+                # verify second transaction is payment
+                Gtxn[1].type_enum() == TxnType.Payment,
+                # verify sender of first transaction is buyback address
+                Gtxn[1].sender() == App.globalGet(Bytes("buyback")),
+                # verify amount of algo received by buyer
+                Gtxn[1].amount() == Minus(Mul(Gtxn[0].asset_amount(), App.globalGet(nominal_price)), Gtxn[1].fee()),
+                Gtxn[1].receiver() == Txn.sender(),
+                # verify maturity date is passed
+                Global.latest_timestamp() > App.globalGet(maturity_date)
+            )
+        ),
+        Return(Int(1))
+    ])
+
+    on_redeem_coupon = Seq([
+        Assert(
+            And(
+                basic_checks,
+                # User sends `B_i` to `DEX_i` lsig.
+                Gtxn[0].type_enum() == TxnType.AssetTransfer,
+                # `DEX_i` sends `B_{i+1}` to the user.
+                Gtxn[1].type_enum() == TxnType.AssetTransfer,
+                Gtxn[0].asset_amount() == Gtxn[1].asset_amount(),
+                # `Dex_i` sends coupon value to user
+                Gtxn[2].type_enum() == TxnType.Payment,
+                Gtxn[2].receiver() == Txn.sender(),
+                # verify coupon amount
+                Gtxn[2].amount() == Mul(Gtxn[0].asset_amount(), App.globalGet(coupon_value))
+            )
+        ),
+        Return(Int(1))
+    ])
+
+    # fetch asset_holding.balance from Txn.accounts[0]
+    asset_balance = AssetHolding.balance(Int(1), Gtxn[2].xfer_asset())
+
+    on_create_dex = Seq([
+        asset_balance, # load asset_balance from store
+        Assert(
+            And(
+                Txn.sender() == App.globalGet(store_manager),
+                basic_checks,
+                Gtxn[1].type_enum() == TxnType.AssetTransfer,
+                # transfer `balanceOf(issuer, B_i)`  of `B_{i+1}` from the creator to the `issuer`.
+                # index 1 of Txn.accounts().
+                asset_balance.value() == Gtxn[1].asset_amount(),
+                # burn `B_i` issuer bonds. send to creator
+                Gtxn[2].type_enum() == TxnType.AssetTransfer,
+                Gtxn[2].asset_receiver() == App.globalGet(creator),
+                asset_balance.value() == Gtxn[2].asset_amount()
+            ),
+        ),
+        # Increment `BondApp.epoch`
+        App.globalPut(epoch, Add(App.globalGet(epoch), Int(1))),
```

Comment:
So just for use-case purposes, we are using `epoch` as a counter here, right? Or is it used this way only?

Reply:
what do you mean ? we have epoch as constant in this example, we have note for this in readme